### PR TITLE
fault injection based on std::thread's

### PR DESCRIPTION
**Pull request by myannyax**

### CODE REVIEW

File: include/yaclib/fault/detail/mutex/shared_timed_mutex.hpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  template <typename _Rep, typename _Period>
+  bool try_lock_shared_for(const std::chrono::duration<_Rep, _Period>& duration) {
+    return try_lock_until(yaclib_std::chrono::steady_clock::now() + duration);
+  }
+  template <typename _Clock, typename _Duration>
+  bool try_lock_shared_until(const std::chrono::time_point<_Clock, _Duration>& duration) {
+    YACLIB_INJECT_FAULT(auto res = _m.try_lock_shared_until(duration));
+
+    if (res) {
+      _shared_mode = true;
```
Comment (MBkkt) [inline include/yaclib/fault/detail/mutex/shared_timed_mutex.hpp:51]: - [ ] Fix data race

File: include/yaclib/log_config.hpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+#ifdef __PRETTY_FUNCTION__
+#define YACLIB_FUNCTION_NAME __PRETTY_FUNCTION__
+#elif defined(__FUNCSIG__)
+#define YACLIB_FUNCTION_NAME __FUNCSIG__
+#elif defined(__func__)
+#define YACLIB_FUNCTION_NAME __func__
+#else
+#define YACLIB_FUNCTION_NAME __FUNCTION__
+#endif
```
Comment (MBkkt) [inline include/yaclib/log_config.hpp:20]:
```suggestion
#ifndef YACLIB_FUNC

[[maybe_unused]] inline void FuncHelper() noexcept {
  #ifdef __PRETTY_FUNCTION__
    #define YACLIB_FUNC 2
  #elif defined(__FUNCSIG__)
    #define YACLIB_FUNC 1
  #else
    #define YACLIB_FUNC 0
  #endif
}

#if YACLIB_FUNC == 2
  #define YACLIB_FUNC __PRETTY_FUNCTION__
#elif YACLIB_FUNC == 1
  #define YACLIB_FUNC __FUNCSIG__
#else
  #define YACLIB_FUNC __func__
#endif

#endif
```

File: include/yaclib/log_config.hpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  do {                                                                                                                 \
+    LogError(cond, message, __FILE__, __LINE__, YACLIB_FUNCTION_NAME);                                                 \
+  } while (false)
+
+#define YACLIB_INFO(cond, message)                                                                                     \
+  do {                                                                                                                 \
+    LogInfo(cond, message, __FILE__, __LINE__, YACLIB_FUNCTION_NAME);                                                  \
+  } while (false)
+
+void LogError(bool condition, std::string_view message, std::string_view file, size_t line, std::string_view function);
```
Comment (MBkkt) [inline include/yaclib/log_config.hpp:32]: check condition in macro

```suggestion
void LogError(std::string_view file, std::size_t line, std::string_view function, std::string_view condition, std::string_view message);
```

File: include/yaclib/log_config.hpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <functional>
+#include <string_view>
+
+using callback_type = std::function<void(std::string_view, std::string_view, size_t, std::string_view)>;
```
Comment (MBkkt) [inline include/yaclib/log_config.hpp:6]:
```suggestion
using LogCallbackType = void(*)(std::string_view, std::string_view, size_t, std::string_view);
```

File: src/executor/thread_factory.cpp [hunk 5]
```diff
@@ -40,7 +40,7 @@ class LightThread final : public IThread {
   }
 
   ~LightThread() final {
-    assert(_thread.get_id() != yaclib_std::this_thread::get_id());
+    YACLIB_ERROR(_thread.get_id() == yaclib_std::this_thread::get_id(), "join itself called");
```
Comment (MBkkt) [inline src/executor/thread_factory.cpp:43]:
```suggestion
    YACLIB_ERROR(_thread.get_id() == yaclib_std::this_thread::get_id(), "Thread try to join itself, probably because you forgot Stop ThreadPool, and ThreadPool dtor was called in ThreadPool thread");
```

File: src/executor/thread_factory.cpp [hunk 6]
```diff
@@ -59,7 +59,7 @@ class HeavyThread final : public IThread {
   void Set(std::size_t priority, std::string_view name, IFuncPtr func, IFuncPtr acquire, IFuncPtr release) {
     {
       std::lock_guard lock{_m};
-      assert(_state == State::Idle);
+      YACLIB_ERROR(_state != State::Idle, "non idle state");
```
Comment (MBkkt) [inline src/executor/thread_factory.cpp:62]:
```suggestion
      YACLIB_ERROR(_state != State::Idle, "Trying run some func on not idle thread");
```

File: src/fault/chrono/system_clock.cpp [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+}
+
+time_t SystemClock::to_time_t(const SystemClock::time_point& time_point) noexcept {
+  return std::chrono::system_clock::to_time_t(time_point);
+}
+
+SystemClock::time_point SystemClock::from_time_t(time_t c_time_point) noexcept {
+  return std::chrono::system_clock::from_time_t(c_time_point);
+}
+
```
Comment (MBkkt) [inline src/fault/chrono/system_clock.cpp:16]: Why do we write this code? it doing nothing
Why can't we use the same implementation for a clock with YACLIB_FAULT=OFF and thread-based fault?

File: src/fault/fault_config.cpp [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+#ifdef YACLIB_FAULT
+  yaclib::detail::Yielder::SetFrequency(freq);
+#endif
+}
+
+void SetSleepTime(uint32_t ns) {
+#ifdef YACLIB_FAULT
+  yaclib::detail::Yielder::SetSleepTime(ns);
+#endif
+}
```
Comment (MBkkt) [inline src/fault/fault_config.cpp:13]: Maybe better make it macro?

File: src/log_config.cpp [hunk 9]
```diff
@@ -0,0 +1,44 @@
+#include <yaclib/log_config.hpp>
+
+static callback_type* g_error_callback = nullptr;
+static callback_type* g_info_callback = nullptr;
+
+void SetErrorCallback(callback_type callback) {
+  auto* func_ref = new callback_type(std::move(callback));
+  auto* prev_callback = g_error_callback;
+  g_error_callback = func_ref;
+  delete prev_callback;
```
Comment (MBkkt) [inline src/log_config.cpp:10]: What is going on here??!?

File: src/log_config.cpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+}
```
Comment (MBkkt) [inline src/log_config.cpp:11]:
```suggestion
void SetErrorCallback(LogCallbackType callback) noexcept {
  gErrorCallback = callback;
}
```

File: src/log_config.cpp [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  auto* func_ref = new callback_type(std::move(callback));
+  auto* prev_callback = g_info_callback;
+  g_info_callback = func_ref;
+  delete prev_callback;
+}
+
+void LogError(bool condition, std::string_view message, std::string_view file, size_t line, std::string_view function) {
+  if (!condition) {
+    return;
+  }
```
Comment (MBkkt) [inline src/log_config.cpp:23]: Check condition out of function defintion

File: test/CMakeLists.txt [hunk 12]
```diff
@@ -27,6 +27,20 @@ set(TEST_UTIL_INCLUDES
...
+#TODO(myannyax) any ways this can be avoided?
+if (YACLIB_FAULT)
+  list(APPEND YACLIB_DEFINITIONS YACLIB_FAULT)
+endif ()
+
+if (YACLIB_LOG STREQUAL "ERROR")
+  list(APPEND YACLIB_DEFINITIONS YACLIB_LOG_ERROR)
+elseif (YACLIB_LOG STREQUAL "INFO")
+  list(APPEND YACLIB_DEFINITIONS YACLIB_LOG_INFO)
+endif()
```
Comment (MBkkt) [inline test/CMakeLists.txt:43]:
```suggestion
if (YACLIB_LOG STREQUAL "ERROR")
  list(APPEND YACLIB_DEFINITIONS YACLIB_LOG_ERROR)
elseif (YACLIB_LOG STREQUAL "INFO")
  list(APPEND YACLIB_DEFINITIONS YACLIB_LOG_ERROR)
  list(APPEND YACLIB_DEFINITIONS YACLIB_LOG_INFO)
endif()
```
?

File: test/test.cpp [hunk 13]
```diff
@@ -13,5 +18,14 @@ int main(int argc, char** argv) {
...
 #endif
   ::testing::InitGoogleTest(&argc, argv);
+  SetErrorCallback([](std::string_view msg, std::string_view file, size_t line, std::string_view function) {
+    GTEST_FATAL_FAILURE_(msg.data());
+  });
+  SetInfoCallback([](std::string_view msg, std::string_view file, size_t line, std::string_view function) {
+    GTEST_COUT << msg << " in" << file << ":" << line << ". Function name: " << function;
+  });
+  SetFrequency(8u);
+  SetSleepTime(200u);
```
Comment (MBkkt) [inline test/test.cpp:28]: Add comment about this constant

File: test/test.cpp [hunk 14]
```diff
@@ -13,5 +18,14 @@ int main(int argc, char** argv) {
...
+  YACLIB_INFO(true, "check YACLIB_INFO");
```
Comment (MBkkt) [inline test/test.cpp:29]: Make logging test and move this there

**Review by MBkkt — CHANGES_REQUESTED** [review]
Also 
* std::size_t instead of size_t, include in <cstddef>
* std::int*_t instead of int*_t, include in <cstdint>
*

File: test/CMakeLists.txt (hunk 12 above)
Comment (myannyax) [reply test/CMakeLists.txt:43]: you meant add info to error one instead of add error to info right?
makes sense, I haven't thought of it 🦔

File: include/yaclib/fault/detail/mutex/shared_timed_mutex.hpp (hunk 1 above)
Comment (myannyax) [reply include/yaclib/fault/detail/mutex/shared_timed_mutex.hpp:51]: why data race? res is true only when mutex is obtained 🤔

File: include/yaclib/log_config.hpp (hunk 3 above)
Comment (myannyax) [reply include/yaclib/log_config.hpp:32]: why arguments reordering?

File: include/yaclib/fault/detail/mutex/shared_timed_mutex.hpp (hunk 1 above)
Comment (MBkkt) [reply include/yaclib/fault/detail/mutex/shared_timed_mutex.hpp:51]: Because its shared lock

File: include/yaclib/log_config.hpp (hunk 4 above)
Comment (myannyax) [reply include/yaclib/log_config.hpp]: I guess with that nothing can be captured inside? Like file stream for file logging?

File: src/fault/fault_config.cpp (hunk 8 above)
Comment (myannyax) [reply src/fault/fault_config.cpp:13]: It seems like I can't simply use ifdef inside macro definition?

File: include/yaclib/fault/detail/condition/condition_variable.hpp [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+  template <typename Predicate>
+  void wait(std::unique_lock<yaclib::detail::Mutex>& lock, Predicate predicate) {
+    auto m = lock.release();
+    auto guard = std::unique_lock<std::mutex>(m->GetImpl(), std::adopt_lock);
+    m->UpdateOwner(yaclib::detail::kInvalidThreadId);
+    YACLIB_INJECT_FAULT(_impl.wait(guard, predicate));
+    guard.release();
+    m->UpdateOwner(yaclib_std::this_thread::get_id());
+    auto new_lock = std::unique_lock(*m, std::adopt_lock);
+    lock.swap(new_lock);
```
[the remark this answers is not in the document]
Comment (myannyax) [reply include/yaclib/fault/detail/condition/condition_variable.hpp]: We need the second update for sure, since if thread comes back from waiting and tries to unlock mutex, It's state can be inconsistent since it was locked inside wait.
It seems that the first one is not needed though since I only have asserts on unlock

[the remark this answers is not in the document]
Comment (MBkkt) [reply include/yaclib/fault/detail/condition/condition_variable.hpp]: Its not about second update, its about this checks

File: include/yaclib/log_config.hpp (hunk 4 above)
Comment (MBkkt) [reply include/yaclib/log_config.hpp]: Its ok, in general they can use static variables, request changes this approach on our side

File: include/yaclib/fault/detail/condition/condition_variable.hpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (MBkkt) [reply include/yaclib/fault/detail/condition/condition_variable.hpp]: I'm about owner checks in overall
Not only in this place

File: include/yaclib/fault/detail/mutex/shared_timed_mutex.hpp (hunk 1 above)
Comment (MBkkt) [reply include/yaclib/fault/detail/mutex/shared_timed_mutex.hpp:51]: You can't unsynchronized write under shared lock

File: src/fault/fault_config.cpp (hunk 8 above)
Comment (MBkkt) [reply src/fault/fault_config.cpp:13]: Yep

File: include/yaclib/fault/detail/condition/condition_variable.hpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (myannyax) [reply include/yaclib/fault/detail/condition/condition_variable.hpp]: well I believe it's not connected to fault injection and I guess whose checks somehow checked in Sanitizers for example so I guess we can omit them

File: src/CMakeLists.txt [hunk 16]
```diff
@@ -27,6 +29,17 @@ if (CORO)
...
+if (YACLIB_FAULT)
+  list(APPEND YACLIB_DEFINITIONS YACLIB_FAULT)
+endif ()
+
+if (YACLIB_LOG STREQUAL "ERROR")
+  list(APPEND YACLIB_DEFINITIONS YACLIB_LOG_ERROR)
+  list(APPEND YACLIB_DEFINITIONS YACLIB_LOG_INFO)
+elseif (YACLIB_LOG STREQUAL "INFO")
+  list(APPEND YACLIB_DEFINITIONS YACLIB_LOG_INFO)
+endif()
```
Comment (MBkkt) [inline src/CMakeLists.txt:41]:
```suggestion
if (YACLIB_LOG STREQUAL "ERROR")
  list(APPEND YACLIB_DEFINITIONS YACLIB_LOG_ERROR)
elseif (YACLIB_LOG STREQUAL "INFO")
  list(APPEND YACLIB_DEFINITIONS YACLIB_LOG_ERROR)
  list(APPEND YACLIB_DEFINITIONS YACLIB_LOG_INFO)
endif()
```

File: include/yaclib/log_config.hpp [hunk 17]
```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <functional>
```
Comment (MBkkt) [inline include/yaclib/log_config.hpp:3]:
```suggestion
```

File: include/yaclib/log_config.hpp [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+#include <string_view>
+
+using callback_type = void (*)(std::string_view file, std::size_t line, std::string_view function,
```
Comment (MBkkt) [inline include/yaclib/log_config.hpp:6]:
```suggestion
using LogСallback = void (*)(std::string_view file, std::size_t line, std::string_view function,
```

File: include/yaclib/log_config.hpp [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+                               std::string_view condition, std::string_view message);
+
+void SetErrorCallback(callback_type callback);
+
+void SetInfoCallback(callback_type callback);
```
Comment (MBkkt) [inline include/yaclib/log_config.hpp:11]: should be in yaclib namespace

File: include/yaclib/log_config.hpp [hunk 20]
```diff
@@ -0,0 +1,50 @@
+#pragma once
```
Comment (MBkkt) [inline include/yaclib/log_config.hpp:1]: should be something like
```C++
#ifdef ERROR_LOG_ON

using callback_type = void (*)(std::string_view file, std::size_t line, std::string_view function,

void SetErrorCallback(callback_type callback);

detail::LogError()

#define YACLIB_ERROR ....

#else

#define YACLIB_ERROR(...) // empty

#enif
```

same for info

Comment (MBkkt) [inline include/yaclib/log_config.hpp:1]: Also I think `yaclib/util/log.hpp` or` yaclib/log.hpp` is better

**Review by MBkkt — CHANGES_REQUESTED** [review]

File: .github/workflows/linux.yml [hunk 21]
```diff
@@ -163,6 +163,8 @@ jobs:
                 -DCMAKE_CXX_COMPILER=${compilers[1]}       \
                 -G"Ninja"                                  \
                 -DYACLIB_LINK_OPTIONS="$link_option"       \
+                -DYACLIB_FAULT=OFF                         \
+                -DYACLIB_LOG="INFO"                        \
```
Comment (MBkkt) [inline .github/workflows/linux.yml:167]: should be ERROR (in test we check some stupid code, that trigger info log)

Comment (myannyax) [reply .github/workflows/linux.yml:167]: but then tests for info logs won't work

Comment (MBkkt) [reply .github/workflows/linux.yml:167]: Maybe we make separate target for test logging?

Comment (myannyax) [reply .github/workflows/linux.yml:167]: well actually they won't be run in that case

File: .github/workflows/linux.yml [hunk 22]
```diff
@@ -133,7 +133,7 @@ jobs:
             "-stdlib=libc++"
             "-stdlib=libstdc++"
           )
-
+
```
Comment (MBkkt) [inline .github/workflows/linux.yml:136]: PlZ remove this unrelated diff in next PR

Comment (MBkkt) [inline .github/workflows/linux.yml:136]: Plz remove this unrelated diff in next PR

File: .github/workflows/linux.yml [hunk 23]
```diff
@@ -150,10 +150,10 @@ jobs:
                   standard=20
                 fi
               fi
-        
+
```
Comment (MBkkt) [inline .github/workflows/linux.yml:153]: Plz remove this unrelated diff in next PR

File: .github/workflows/linux.yml [hunk 24]
```diff
@@ -150,10 +150,10 @@ jobs:
                   standard=20
                 fi
               fi
-        
+          
               dir="build_${compilers[0]}_${stdlibs_name[$j]}_$build_type"
               echo $dir
-
+
```
Comment (MBkkt) [inline .github/workflows/linux.yml:156]: Plz remove this unrelated diff in next PR

File: .github/workflows/linux.yml [hunk 25]
```diff
@@ -163,6 +163,8 @@ jobs:
                 -DCMAKE_CXX_COMPILER=${compilers[1]}       \
                 -G"Ninja"                                  \
                 -DYACLIB_LINK_OPTIONS="$link_option"       \
+                -DYACLIB_FAULT=OFF                         \
+                -DYACLIB_LOG="ERROR"                       \
```
Comment (MBkkt) [inline .github/workflows/linux.yml:167]: What do you think about disable/enable error check in release depends on some `rand`?

Comment (MBkkt) [reply .github/workflows/linux.yml:167]: Same for YACLIB_FAULT

File: .github/workflows/windows.yml [hunk 26]
```diff
@@ -85,17 +85,18 @@ jobs:
           options=("-A $arch -T ClangCL" "-DCMAKE_BUILD_TYPE=${{ matrix.build_type }}" "-A $arch")
 
           for (( i=0; i<${#names[*]}; i+=1 )); do
-            dir="build_${names[$i]}"
+            dir="build_${names[$i]}_fault_${yaclib_fault}"
             echo $dir
-
+
```
Comment (MBkkt) [inline .github/workflows/windows.yml:90]: Plz remove this unrelated diff in next PR

File: include/yaclib/fault/detail/mutex/mutex.hpp [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+  Mutex(const Mutex&) = delete;
+  Mutex& operator=(const Mutex&) = delete;
+
+  void lock();
+  bool try_lock() noexcept;
+  void unlock() noexcept;
+
+  using native_handle_type = std::mutex::native_handle_type;
+
+  inline native_handle_type native_handle();
```
Comment (MBkkt) [inline include/yaclib/fault/detail/mutex/mutex.hpp:25]: unnecessary inline

File: include/yaclib/fault/detail/mutex/recursive_timed_mutex.hpp [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+  RecursiveTimedMutex(const RecursiveTimedMutex&) = delete;
+  RecursiveTimedMutex& operator=(const RecursiveTimedMutex&) = delete;
+
+  void lock();
+  bool try_lock() noexcept;
+  void unlock() noexcept;
+
+  template <typename _Rep, typename _Period>
+  bool try_lock_for(const std::chrono::duration<_Rep, _Period>& duration) {
+    return try_lock_until(yaclib_std::chrono::steady_clock::now() + duration);
```
Comment (MBkkt) [inline include/yaclib/fault/detail/mutex/recursive_timed_mutex.hpp:25]: You should use native methods

File: include/yaclib/fault/detail/mutex/shared_timed_mutex.hpp [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+  SharedTimedMutex(const SharedTimedMutex&) = delete;
+  SharedTimedMutex& operator=(const SharedTimedMutex&) = delete;
+
+  void lock();
+  bool try_lock() noexcept;
+  void unlock() noexcept;
+
+  template <typename _Rep, typename _Period>
+  bool try_lock_for(const std::chrono::duration<_Rep, _Period>& duration) {
+    return try_lock_until(yaclib_std::chrono::steady_clock::now() + duration);
```
Comment (MBkkt) [inline include/yaclib/fault/detail/mutex/shared_timed_mutex.hpp:25]: same

File: include/yaclib/fault/detail/mutex/timed_mutex.hpp [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+  TimedMutex(const TimedMutex&) = delete;
+  TimedMutex& operator=(const TimedMutex&) = delete;
+
+  void lock();
+  bool try_lock() noexcept;
+  void unlock() noexcept;
+
+  template <typename _Rep, typename _Period>
+  bool try_lock_for(const std::chrono::duration<_Rep, _Period>& duration) {
+    return try_lock_until(yaclib_std::chrono::steady_clock::now() + duration);
```
Comment (MBkkt) [inline include/yaclib/fault/detail/mutex/timed_mutex.hpp:26]: same

**Review by MBkkt — APPROVED** [review]
looks good enough for merge, but you should create additional PR with fixes

Also you forgot make beautiful commits history
